core/pubnub_coreapi.c

```diff
@@ -281,22 +281,141 @@ enum pubnub_res pubnub_set_state(pubnub_t*   pb,
+                    size_t ch_len;
+                    bool end = false;
+                    do{
+                        ch_temp = strchr(str_ch,',');
+                        if (ch_cnt > 0) { 
+                            memcpy(json_state + m_len, ",", cm_len);
+                            m_len += cm_len;
+                        }
+                        if (ch_temp == NULL) {
+                            end = true;
+                            ch_len = strlen(str_ch);
+                        }
+                        else { ch_len = ch_temp - str_ch; }
+
+                        if (ch_len == 0) { continue; }
+
+                        char curr_ch[62];
+                        strncpy(curr_ch, str_ch, ch_len);
+                        curr_ch[ch_len] = '\0';
+
+                        memcpy(json_state + m_len, "\"", dq_len);
+                        m_len += dq_len;
+                        memcpy(json_state + m_len, curr_ch, ch_len);
+                        m_len += ch_len;
+                        memcpy(json_state + m_len, "\"", dq_len);
+                        m_len += dq_len;
+                        memcpy(json_state + m_len, ":", co_len);
+                        m_len += co_len;
+                        memcpy(json_state + m_len, state, st_len);
+                        m_len += st_len;
+
+                        ch_cnt++;
+
+                        str_ch = ch_temp + 1;
+                    } while (false == end);
+                }
+                
+                if (channel_group) {
+                    char* str_cg = (char*)channel_group;
+                    char* cg_temp;
+                    int cg_len;
+                    bool end = false;
+                    do{
+                        cg_temp = strchr(str_cg,',');
+                        if (ch_cnt > 0 || cg_cnt > 0) { 
+                            memcpy(json_state + m_len, ",", cm_len);
+                            m_len += cm_len;
+                        }
+                        if (cg_temp == NULL) {
+                            end = true;
+                            cg_len = strlen(str_cg);
+                        }
+                        else { cg_len = cg_temp - str_cg; }
+
+                         if (cg_len == 0) { continue; }
+
+                        char curr_cg[62];
+                        strncpy(curr_cg, str_cg, cg_len);
+                        curr_cg[cg_len] = '\0';
+
+                        memcpy(json_state + m_len, "\"", dq_len);
+                        m_len += dq_len;
+                        memcpy(json_state + m_len, curr_cg, cg_len);
+                        m_len += cg_len;
+                        memcpy(json_state + m_len, "\"", dq_len);
+                        m_len += dq_len;
+                        memcpy(json_state + m_len, ":", co_len);
+                        m_len += co_len;
+                        memcpy(json_state + m_len, state, st_len);
+                        m_len += st_len;
```

Comment:
In the past, I tried to do sprintf to do simple oneliner, but address sanizer is throwing heap memory leak issues. So I had to rewrite with memcpy, which passed sanitizer test.

Reply:
exatly I tried like this, but I am not getting the dictionary format I want.  Memory addresses are getting overwritten with both channel and channel group. Final output is weird result.

Reply:
Actually your code little different from mine.
we need previous "added_len" value before calling this function. Then we increment the memory location to add chars. Then return the memory location/position or length back to the caller, for next channel or channel group.
Having internal  variable added_len =0, would not give expected result.
Since I am concatenating chars to json, we need to increment the memory position, else won't get expected result.

Reply:
ok..created `int json_kvp_builder(char* jsonbuilder, int pos, char* key, char* val)`   which can be reused, based on your comment.